### PR TITLE
strengthen-find-path

### PR DESCRIPTION
**Pull request by Amber1990Zhang**
https://github.com/vesoft-inc/nebula-common/pull/358
https://github.com/vesoft-inc/nebula-graph/pull/459

### CODE REVIEW

File: docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+<vertex_id_list> ::=
+    [vertex_id [, vertex_id ...]
+```
+
+The `FIND PATH` statement finds the paths between the selected source vertex and destination vertex.
+
+- `SHORTEST` finds the shortest path.
+- `ALL` finds all the paths.
+- `<vertex_id_list>` is a list of vertex IDs separated by commas (,). It supports `$-` and `$var`.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:15]: with commas

File: docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+- `<edge_type_list>`  is a list of edge types separated by commas (,). `*` is all edge types.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:16]: with commas

Comment (oldLady344) [reply docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:16]: `*` indicates all edge types.

File: docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+- `<N>` is the hop number. The default value is 5.
+
+## Limitations
+
+- When source and destination vertices are ID lists, the shortest path between any source vertices and the destination vertices is returned.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:21]: "When source and destination vertices are ID lists" ==> When a list of source and/or destination vertex IDs are specified"?

Comment (Amber1990Zhang) [reply docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:21]: nice!

File: docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+- `<vertex_id_list>` is a list of vertex IDs separated with commas (,). It supports `$-` and `$var`.
+- `<edge_type_list>`  is a list of edge types separated with commas (,). `*` is all edge types.
+- `<N>` is the hop number. The default value is 5.
+
+## Limitations
+
+- When a list of source and/or destination vertex IDs are specified, the shortest path between any source vertices and the destination vertices is returned.
+- There can be cycles when searching all paths.
+- `FIND PATH` does not support property filtering.
+- `FIND PATH` does not support specifying direction.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:24]: specifying a direction?

File: docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+- `SHORTEST` finds the shortest path.
+- `ALL` finds all the paths.
+- `<vertex_id_list>` is a list of vertex IDs separated with commas (,). It supports `$-` and `$var`.
+- `<edge_type_list>`  is a list of edge types separated with commas (,). `*` is all edge types.
+- `<N>` is the hop number. The default value is 5.
+
+## Limitations
+
+- When a list of source and/or destination vertex IDs are specified, the shortest path between any source vertices and the destination vertices is returned.
+- There can be cycles when searching all paths.
```
Comment (whitewum) [inline docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:22]: combine this item with `When finding path with the `NOLOOP` keyword, the returned paths do not include cycles.`

If `NOLOOP` is not used, There can be cycles when searching all paths. If `NOLOOP` is used, the returned paths do not include cycles

**Review by whitewum — COMMENTED** [review]
minor

File: docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md (hunk 5 above)
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:22]: `FIND PATH` can retrieve a graph containing cycles?

"there can be" is not popular. And always avoid "there be" in technical writing.

File: docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+- `FIND PATH` does not support property filtering.
+- `FIND PATH` does not support specifying direction.
+- `FIND PATH` is a single-process procedure, so it uses much memory.
+- When finding path with the `NOLOOP` keyword, the returned paths do not include cycles.
+
+## Examples
+
+In Nebula Console, path is shown as `vertex_id <edge_name, rank> vertex_id`.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:30]: a path?

File: docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+| path                                                             |
++------------------------------------------------------------------+
+| ("player102")-[:follow@0]->("player101")-[:serve@0]->("team201") |
++------------------------------------------------------------------+
+```
+
+```ngql
+nebula> FIND SHORTEST PATH FROM "player100" TO "team200" OVER * REVERSELY;
+Empty set (time spent 17442/18406 us)
+```
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:44]: should we give some explanation about the "Empty set" result?

File: docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+## Limitations
+
+- When a list of source and/or destination vertex IDs are specified, the shortest path between any source vertices and the destination vertices is returned.
+- There can be cycles when searching all paths.
+- `FIND PATH` does not support property filtering.
+- `FIND PATH` does not support specifying direction.
+- `FIND PATH` is a single-process procedure, so it uses much memory.
+- When finding path with the `NOLOOP` keyword, the returned paths do not include cycles.
+
+## Examples
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:28]: should we give an example of "NOLOOP"?

File: docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md (hunk 7 above)
Comment (Amber1990Zhang) [reply docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:44]: I’m not sure. But if it's necessary I don’t think here is a good idea

File: docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md [hunk 9]
```diff
@@ -1,33 +1,35 @@
 # FIND PATH
 
 ```ngql
-FIND SHORTEST | ALL | NOLOOP PATH FROM <vertex_id_list> TO <vertex_id_list>
-OVER <edge_type_list> [REVERSELY] [BIDIRECT] [UPTO <N> STEPS]
+FIND { SHORTEST | ALL | NOLOOP } PATH FROM <vertex_id_list> TO <vertex_id_list>
+OVER <edge_type_list> [REVERSELY | BIDIRECT] [UPTO <N> STEPS] [| ORDER BY $-.path] [| LIMIT <M>]
```
Comment (whitewum) [inline docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:5]: Great!

**Review by whitewum — DISMISSED** [review]

File: docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md [hunk 10]
```diff
@@ -1,33 +1,35 @@
...
 ## Limitations
 
 - When a list of source and/or destination vertex IDs are specified, the shortest path between any source vertices and the destination vertices is returned.
 - There can be cycles when searching all paths.
 - `FIND PATH` does not support property filtering.
-- `FIND PATH` does not support specifying direction.
+- `FIND PATH` does not support specifying a direction.
 - `FIND PATH` is a single-process procedure, so it uses much memory.
 - When finding path with the `NOLOOP` keyword, the returned paths do not include cycles.
+- If `NOLOOP` is not used, `FIND PATH` can retrieve paths contains cycles. If `NOLOOP` is used, `FIND PATH` can retrieve paths without cycles.
```
Comment (oldLady344) [inline docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:28]: containing

File: docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md (hunk 7 above)
Comment (oldLady344) [reply docs-2.0/3.ngql-guide/16.subgraph-and-path/2.find-path.md:46]: > I’m not sure. But if it's necessary I don’t think here is a good idea

Why do you give such an example. When I read an example that gives an empty result, I suppose you have some special purpose. SO I have the preceding question.

**Review by whitewum — APPROVED** [review]

**Review by randomJoe211 — APPROVED** [review]
👍

**Pull request merged**